sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RemoveRedundantAggregates.scala

```diff
@@ -42,11 +42,12 @@ object RemoveRedundantAggregates extends Rule[LogicalPlan] with AliasHelper {
       )
 
       // We might have introduces non-deterministic grouping expression
```

Comment:
I have a different idea. I think it's risky to put the new grouping expressions back to the `Aggregate`, as the grouping expression contains things in the SELECT list. This is a long-standing issue and I feel it's better to just create a `Project` below the `Aggregate` to calculate grouping expressions, and other optimizer rules can merge/eliminate this extra `Project` if it only contains `Attributes`.

Reply:
@cloud-fan Thanks. I'm sorry I didn't understand. Before this PR, it will create a `Project` below the `Aggregate` to Pulls out nondeterministic grouping expressions. In this PR, I keep this logic, and create a `Project` below the `Aggregate` to Pulls out complex grouping expressions by rule `PullOutGroupingExpressions`. I do not put the new grouping expressions back to the `Aggregate`.

Reply:
let me put it this way: if we can safely remove a distinct-like `Aggregate`, it means we can just turn it into a `Project` (e.g. `case a: Aggregate => Project(a.aggregateExpressions, a.child)`).

The code today is merging the distinct-like `Aggregate` to the upper `Aggregate`, which is problematic and we have to pull something out to a `Project` case by case. This is fragile and this PR just finds another case.

My proposal is to not merge. Just convert the distinct-like `Aggregate` to `Project`, and let other optimizer rules decide if we can merge it or not.

Reply:
Thank you, I understand, I will resubmit a PR later.